### PR TITLE
Make Languages Dynamic and Add Language Manager

### LINKED ISSUE #501: make languages dynamic
- make a languages table schema similar to skills
- seed new table with languages enum
- create new a `createLanguage`, `languages`, and  `deleteLanguage` resolver
- update all types and usage across backend
- update usage across frontend 

Refer to https://github.com/uwblueprint/sistering/pull/482/files for frontend queries

update languages field to search field that creates new skills if not exist, similar to #502
Part of account creation flow. We should change the languages dropdown to a search bar that also creates new skills if they do not exists (trimmed)

### PR DESCRIPTION
**Pull request by briantu**
## Ticket link

Closes #501, #544 

## Implementation description
* Created table, types, and resolvers for languages in backend
* Created language manager in frontend

## Steps to test
1. Go to `http://localhost:3000/admin`
2. Click on gear icon
3. Click on languages tab
4. Test CRUD operations on languages

## What should reviewers focus on?
* Code correctness
* Design correctness

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Review by MatoPlus — COMMENTED** [review]
Note to self: see #560 was fixed as a side effect in this PR

Edit: Cannot reproduce the bug in this branch. We can probably close #560 as well

**Review by MatoPlus — COMMENTED** [review]
Good seems good with many nice cleanups as well. Will test and approve in a bit : )

**Review by MatoPlus — APPROVED** [review]
LGTM, awesome work 💪 

I think this is fine to merge but other reviews could be nice @sherryhli @neelismail01

File: backend/services/implementations/languageService.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    return {
+      id: String(language.id),
+      name: language.name,
+    };
+  }
+
+  async getLanguages(): Promise<LanguageResponseDTO[]> {
+    try {
+      const languages: Array<Language> = await prisma.language.findMany({
+        orderBy: [{ id: "asc" }],
```
Comment (sherryhli) [inline backend/services/implementations/languageService.ts:43]: Why do we require sorting and why by `id`? If it's to help users when they make selections then maybe sorting by name would be more helpful :)

File: backend/services/implementations/userService.ts [hunk 2]
```diff
@@ -49,20 +50,34 @@ export const convertToSkillResponseDTO = (
   });
 };
 
+export const convertToLanguageResponseDTO = (
+  languages: Language[],
+): LanguageResponseDTO[] => {
+  return languages.map((language: Skill) => {
```
Comment (sherryhli) [inline backend/services/implementations/userService.ts:56]:
```suggestion
  return languages.map((language: Language) => {
```

File: backend/types.ts [hunk 3]
```diff
@@ -75,6 +67,7 @@ export type VolunteerUserRequestDTO = UserDTO &
   Omit<VolunteerDTO, "skills" | "branches"> & {
     skills: string[];
     branches: string[];
+    languages: string[];
```
Comment (sherryhli) [inline backend/types.ts:70]: I think `languages` needs to be omitted from `UserDTO` above to prevent a typing error

File: backend/types.ts [hunk 4]
```diff
@@ -93,6 +86,7 @@ export type EmployeeUserDTO = {
 export type EmployeeUserRequestDTO = UserDTO &
   EmployeeUserDTO & {
     branches: string[];
+    languages: string[];
```
Comment (sherryhli) [inline backend/types.ts:89]: Same thing here

File: frontend/src/components/pages/admin/user/AdminUserManagementPage.tsx [hunk 5]
```diff
@@ -345,7 +350,7 @@ const AdminUserManagementPage = (): React.ReactElement => {
                 rowEmgEmail = user.emergencyContactEmail ?? "N/A";
                 rowEmgName = user.emergencyContactName ?? "N/A";
                 rowEmgPhone = user.emergencyContactPhone ?? "N/A";
-                rowLangs = user.languages;
+                rowLangs = user.languages?.map((language) => language.name);
```
Comment (sherryhli) [inline frontend/src/components/pages/admin/user/AdminUserManagementPage.tsx:353]: Can `user.languages` ever be null?

**Review by sherryhli — COMMENTED** [review]
Quite a hefty PR, nice work! Would suggest addressing these non-critical issues before merging.

To confirm, the search bar feature is not being implemented right?
> update languages field to search field that creates new skills if not exist, similar to https://github.com/uwblueprint/sistering/issues/502
Part of account creation flow. We should change the languages dropdown to a search bar that also creates new skills if they do not exists

File: backend/services/implementations/languageService.ts (hunk 1 above)
Comment (MatoPlus) [reply backend/services/implementations/languageService.ts:43]: by default it sorts by updatedAt, this causes an unwanted rendering order switch on the language manager when we fetch data. I like the suggestion to sort by name tho, In which case we also want to do the same for branches and skills

Comment (briantu) [reply backend/services/implementations/languageService.ts:43]: Sure, I'll change it to sort by name

File: frontend/src/components/pages/admin/user/AdminUserManagementPage.tsx (hunk 5 above)
Comment (briantu) [reply frontend/src/components/pages/admin/user/AdminUserManagementPage.tsx]: Oh I suppose not. I was just following what we had for skills 😅. I don't think `user.skills` will ever be null either right?
https://github.com/uwblueprint/sistering/blob/6f5e48ffdbf93667db479630d3f5f02d18a0ee27/frontend/src/components/pages/admin/user/AdminUserManagementPage.tsx#L354

Comment (sherryhli) [reply frontend/src/components/pages/admin/user/AdminUserManagementPage.tsx]: Yeah, don't think it ever will

**Review by sherryhli — APPROVED** [review]
LGTM 🚢

**Comment by MatoPlus** [comment]
> Quite a hefty PR, nice work! Would suggest addressing these non-critical issues before merging.
> 
> To confirm, the search bar feature is not being implemented right?
> 
> > update languages field to search field that creates new skills if not exist, similar to #502
> > Part of account creation flow. We should change the languages dropdown to a search bar that also creates new skills if they do not exists

The search item you talked about is not being implemented in this PR, it may be a follow up if we successfully create the skills variant that @twilkhoo is working on.

**Pull request merged**